src/jailer/src/env.rs

```diff
@@ -168,20 +171,34 @@ impl Env {
 
         SyscallReturnCode(unsafe { libc::chown(dev_path.as_ptr(), self.uid(), self.gid()) })
             .into_empty_result()
-            .map_err(|e| Error::ChangeFileOwner(dev_path.to_str().unwrap(), e))
+            // Safe to unwrap as we provided valid file names.
+            .map_err(|e| Error::ChangeFileOwner(PathBuf::from(dev_path.to_str().unwrap()), e))
     }
 
-    fn setup_jailed_folders(&mut self) -> Result<()> {
-        for folder in FOLDER_HIERARCHY.iter() {
-            let folder_cstr = CStr::from_bytes_with_nul(folder).map_err(Error::FromBytesWithNul)?;
-
-            // This unwrap is safe as we provided strings that have valid utf8 chars.
+    fn setup_jailed_folders<P>(&self, folders: Vec<P>) -> Result<()>
+    where
+        P: AsRef<Path> + AsRef<OsStr>,
+    {
+        for folder in folders.iter() {
+            let mut folder_bytes = Path::new(folder).as_os_str().as_bytes().to_vec();
+            folder_bytes.push(b'\0');
+            let folder_cstr = CStr::from_bytes_with_nul(folder_bytes.as_slice())
+                .map_err(Error::FromBytesWithNul)?;
+
+            // Safe to unwrap as the byte sequence is UTF-8 validated above.
             let path = folder_cstr.to_str().unwrap();
-            fs::create_dir_all(path).map_err(|e| Error::CreateDir(PathBuf::from(path), e))?;
-
-            SyscallReturnCode(unsafe { libc::chown(folder_cstr.as_ptr(), self.uid(), self.gid()) })
+            let path_buf = PathBuf::from(path);
+            fs::create_dir_all(path).map_err(|e| Error::CreateDir(path_buf.clone(), e))?;
+            fs::set_permissions(path, Permissions::from_mode(FOLDER_PERMISSIONS))
```

Comment:
Doesn't this mean that each jail root is now accessible by any user on the system (including `uid`s of other firecracker instances? I would avoid this. Why isn't the `chown` enough?

Reply:
After `chown`-ing the file/dir, I expect `0o700` (`rwx|___|___`) to be just right so only `uid` (and root) can access the resources in its jail.

Reply:
You're right that only the owner bits should be set. `chown` isn't enough in cases where a preconfigured `umask` removes access rights on the directory. I'll change to `700`.